core/src/main/java/org/infinispan/util/concurrent/locks/impl/DefaultPendingLockManager.java

```diff
@@ -288,17 +296,77 @@ private void forEachTransaction(Consumer<CacheTransaction> consumer) {
       }
    }
 
+   private static long awaitOn(DefaultPendingLockPromise pendingLockPromise, GlobalTransaction globalTransaction)
+         throws InterruptedException {
+      pendingLockPromise.await();
+      if (pendingLockPromise.hasTimedOut()) {
+         timeout(pendingLockPromise.getTimedOutTransaction(), globalTransaction);
+      }
+      return pendingLockPromise.getRemainingTimeout();
+   }
+
    private static class PendingTransaction {
       public final CacheTransaction cacheTransaction;
-      public final Object commonKey;
+      public final Object key;
+      private final CompletableFuture<Void> keyReleased;
 
-      private PendingTransaction(CacheTransaction cacheTransaction, Object commonKey) {
+      private PendingTransaction(CacheTransaction cacheTransaction, Object key, CompletableFuture<Void> keyReleased) {
          this.cacheTransaction = cacheTransaction;
-         this.commonKey = commonKey;
+         this.key = key;
+         this.keyReleased = Objects.requireNonNull(keyReleased);
+      }
+   }
+
+   private interface DefaultPendingLockPromise extends PendingLockPromise, Callable<Void> {
+
+      void await() throws InterruptedException;
+
+      PendingTransaction getTimedOutTransaction();
```

Comment:
this is only implementation specific method. I don't wanna make them public outside this class.